tests/python/pants_test/option/test_enclosing_scopes.py

```diff
@@ -0,0 +1,66 @@
+# coding=utf-8
+# Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from pants.option.arg_splitter import GLOBAL_SCOPE
+from pants.option.parser_hierarchy import InvalidScopeError, all_enclosing_scopes, enclosing_scope
+from pants_test.base_test import BaseTest
+
+
+class TestEnclosingScopeTraversal(BaseTest):
+  def test_enclosing_scope(self):
+    """The enclosing scope of any non-nested scope should be the global scope,
+    and the enclosing scope of a nested scope should be the scope without its
+    final component."""
+    self.assertEqual(GLOBAL_SCOPE, enclosing_scope(GLOBAL_SCOPE))
+    self.assertEqual(GLOBAL_SCOPE, enclosing_scope('scope'))
+    self.assertEqual('base', enclosing_scope('base.subscope'))
+
+  def test_all_enclosing_scopes(self):
+    """`all_enclosing_scopes` should repeatedly apply `enclosing_scope` to any
+    valid single- or multiple- component scope. `all_enclosing_scopes` should
+    not yield the global scope if `allow_global=False`."""
+    global_closure = list(all_enclosing_scopes(GLOBAL_SCOPE, allow_global=True))
+    self.assertEqual(global_closure, [GLOBAL_SCOPE])
+
+    global_closure_excluded = list(all_enclosing_scopes(GLOBAL_SCOPE, allow_global=False))
+    self.assertEqual(global_closure_excluded, [])
+
+    base_scope = 'scope'
+    base_scope_closure = list(all_enclosing_scopes(base_scope))
+    self.assertEqual(base_scope_closure, [base_scope, GLOBAL_SCOPE])
+
+    subscope = 'subscope'
+    compound_scope = '{}.{}'.format(base_scope, subscope)
+    compound_scope_closure = list(all_enclosing_scopes(compound_scope))
+    self.assertEqual(compound_scope_closure, [compound_scope, base_scope, GLOBAL_SCOPE])
```

Comment:
seems the order of return value from `all_enclosing_scopes` matters, could you add it to the doc that it will return in the order from most deepest subscope to global scope?